pyvrp/Population.py

```diff
@@ -398,3 +564,51 @@ def get_best_found(self) -> Individual:
         """
         # TODO move this best solution stuff to GeneticAlgorithm
         return self._best
+
+
+_IndividualDiversityMeasure = Callable[
+    [ProblemData, Individual, Individual], float
+]
+
+
+class Population(GenericPopulation[Individual]):
+    def __init__(
+        self,
+        data: ProblemData,
+        penalty_manager: PenaltyManager,
+        rng: XorShift128,
+        diversity_op: _IndividualDiversityMeasure,
+        initial_solutions: Optional[List[Individual]] = None,
+        params: PopulationParams = PopulationParams(),
+    ):
+        """
+        Creates a Population instance.
+
+        Parameters
+        ----------
+        data
+            Data object describing the problem to be solved.
+        penalty_manager
+            Penalty manager to use.
+        rng
+            Random number generator.
+        diversity_op
+            Operator to use to determine pairwise diversity between solutions.
+        initial_solutions
+            List of Individuals to populate the population with.
+            Will be generated randomly if None.
+        params, optional
+            Population parameters. If not provided, a default will be used.
+        """
+        super().__init__(
+            Individual(data, penalty_manager, rng),
+            lambda first, second: diversity_op(data, first, second),
+            rng.randint,
+            params,
+        )
+
+        if initial_solutions:
+            for indiv in initial_solutions:
+                self.add(indiv)
+        else:
+            self.initialize(lambda: Individual(data, penalty_manager, rng))
```

Comment:
After going through this class, I am a little overwhelmed. So much of this exists just to satisfy the type checker, when I do not immediately see the benefit w.r.t. the code that we actually have, rather than what we might write in a few months.

I see the value of having this work for general types that satisfy some constraints (e.g., have an objective value), but I question the relatively heavy-handed use of classes and subclasses here. Is there a way to do this with less work? Do we need to branch on numpy/pure python in the type system, for example? 

(I don't know, but that's a concern I have)